skyportal/handlers/api/observation.py

```diff
@@ -0,0 +1,488 @@
+from baselayer.app.access import permissions, auth_or_token
+from baselayer.log import make_log
+import arrow
+import healpix_alchemy as ha
+import pandas as pd
+import sqlalchemy as sa
+from sqlalchemy.orm import joinedload
+from sqlalchemy.orm import sessionmaker, scoped_session
+from tornado.ioloop import IOLoop
+from astropy.time import Time
+
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    GcnEvent,
+    Localization,
+    LocalizationTile,
+    Telescope,
+    Instrument,
+    InstrumentField,
+    InstrumentFieldTile,
+    ExecutedObservation,
+)
+
+
+log = make_log('api/observation')
+
+Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
+
+
+def add_observations(instrument_id, obstable):
+    """Post executed observations for a given instrument.
+    obstable is a pandas DataFrame of the form:
+
+   observation_id  field_id       obstime   seeing    limmag  exposure_time  \
+0        84434604         1  2.458599e+06  1.57415  20.40705             30
+1        84434651         1  2.458599e+06  1.58120  20.49405             30
+2        84434696         1  2.458599e+06  1.64995  20.56030             30
+3        84434741         1  2.458599e+06  1.54945  20.57400             30
+4        84434788         1  2.458599e+06  1.62870  20.60385             30
+
+  filter  processed_fraction airmass
+0   ztfr                 1.0    None
+1   ztfr                 1.0    None
+2   ztfr                 1.0    None
+3   ztfr                 1.0    None
+4   ztfr                 1.0    None
+     """
+
+    session = Session()
+    try:
+        observations = []
+        for index, row in obstable.iterrows():
+            field_id = int(row["field_id"])
+            field = (
+                session.query(InstrumentField)
+                .filter_by(instrument_id=instrument_id, field_id=field_id)
+                .first()
+            )
+            if field is None:
+                return log(
+                    f"Unable to add observations for instrument {instrument_id}: Missing field {field_id}"
+                )
+
+            observation = (
+                session.query(ExecutedObservation)
+                .filter_by(
+                    instrument_id=instrument_id, observation_id=row["observation_id"]
+                )
+                .first()
+            )
+            if observation is not None:
+                log(
+                    f"Observation {row['observation_id']} for instrument {instrument_id} already exists... continuing."
```

Comment:
It already exists... but does this check prevent adding a duplicate?

Reply:
Added a continue :)